src/history/sqlite_backed.rs

```diff
@@ -0,0 +1,410 @@
+            .map(|e| Duration::from_millis(e as u64)),
+        exit_status: row.get("exit_status")?,
+        more_info: x
+            .map(|x| {
+                serde_json::from_str(&x).map_err(|e| {
+                    // hack
+                    rusqlite::Error::InvalidColumnType(
+                        0,
+                        format!("could not deserialize more_info: {e}"),
+                        rusqlite::types::Type::Text,
+                    )
+                })
+            })
+            .transpose()?,
+    })
+}
+
+impl History for SqliteBackedHistory {
+    fn save(&mut self, mut entry: HistoryItem) -> Result<HistoryItem> {
+        /*if entry.id.is_some() {
+            return Err("ID must be empty".to_string());
+        }*/
+        let ret: i64 = self
+            .db
+            .prepare(
+                "insert into history
+                               (id,  start_timestamp,  command_line,  session_id,  hostname,  cwd,  duration_ms,  exit_status,  more_info)
+                        values (:id, :start_timestamp, :command_line, :session_id, :hostname, :cwd, :duration_ms, :exit_status, :more_info)
+                    on conflict (history.id) do update set
+                        start_timestamp = excluded.start_timestamp,
+                        command_line = excluded.command_line,
+                        session_id = excluded.session_id,
+                        hostname = excluded.hostname,
+                        cwd = excluded.cwd,
+                        duration_ms = excluded.duration_ms,
+                        exit_status = excluded.exit_status,
+                        more_info = excluded.more_info
+                    returning id",
+            )
+            .map_err(|e|e.to_string())?
+            .query_row(
+                named_params! {
+                    ":id": entry.id.map(|id| id.0),
+                    ":start_timestamp": entry.start_timestamp.map(|e| e.timestamp_millis()),
+                    ":command_line": entry.command_line,
+                    ":session_id": entry.session_id.map(|e| e.0),
+                    ":hostname": entry.hostname,
+                    ":cwd": entry.cwd,
+                    ":duration_ms": entry.duration.map(|e| e.as_millis() as i64),
+                    ":exit_status": entry.exit_status,
+                    ":more_info": entry.more_info.as_ref().map(|e| serde_json::to_string(e).unwrap())
+                },
+                |row| row.get(0),
+            )
+            .map_err(|e| e.to_string())?;
+        entry.id = Some(HistoryItemId::new(ret));
+        Ok(entry)
+    }
+
+    fn load(&self, id: HistoryItemId) -> Result<HistoryItem> {
+        let entry = self
+            .db
+            .prepare("select * from history where id = :id")
+            .map_err(|e| e.to_string())?
+            .query_row(named_params! { ":id": id.0 }, deserialize_history_item)
+            .map_err(|e| e.to_string())?;
```

Comment:
The stringly typed `Err`s should go away. What failure cases can we imagine that are handleable by the downstream code? Schema mismatch as a result of versioning, file inaccessibility (basically the std::io::Error). What else? Feature that is not supported. (also applies to the simpler file/in memory impl)